code/go/0chain.net/smartcontract/dbs/event/miner.go

```diff
@@ -30,26 +32,54 @@ type Miner struct {
 	Rewards           state.Balance
 	Fees              state.Balance
 	Active            bool
-	Longitude         int64
-	Latitude          int64
+	Longitude         float64
+	Latitude          float64
 }
 
-func (edb *EventDb) GetMiner(id string) (*Miner, error) {
+func (edb *EventDb) GetMiner(id string) (Miner, error) {
```

Comment:
any reason to not use reference?

Reply:
returning a pointer sends the data to heap instead of the stack. 
There was no reason that we want to return a pointer here.

The less we send objects to stack the faster the program because less work for the garbage collector and in turn less garbage collector pauses in golang.

Reply:
Note: garbage collector pauses after the heap allocation doubles or after a specific time whichever is earlier.

Reply:
@samtholiya we should not have a hard and fast rule here.
These structs are big structs and doesn't make sense to return copy every time.

We should be smart in when to return by value and pointer.
If returning pointer would be so bad, it would've not been given by nature of language to us.
the gc optimisation you'll be getting from this won't be too much, but creating copies of these structs every time is expensive. So we have to weigh the trade-offs

For big structs, I would not recommend to return value, but pointer.

Reply:
okay... but this is not a big struct... And slices/maps that can be considered as bigger structs are by default reference driven.
Pointer return is allowed by golang for `mutex` or `struct containing mutex` because if you are returning a mutex it should be a pointer. 
or 
The pointer should be used in places where you want to achieve something similar to `flag` 

But otherwise using a pointer like the json.Unmarshal or marshal uses actually is better

Reply:
@samtholiya it is a big struct, quite evidently:
```json
type Miner struct {
	gorm.Model
	MinerID           string `gorm:"uniqueIndex"`
	N2NHost           string `gorm:"column:n2n_host"`
	Host              string
	Port              int
	Path              string
	PublicKey         string
	ShortName         string
	BuildTag          string
	TotalStaked       state.Balance
	Delete            bool
	DelegateWallet    string
	ServiceCharge     float64
	NumberOfDelegates int
	MinStake          state.Balance
	MaxStake          state.Balance
	LastHealthCheck   common.Timestamp
	Rewards           state.Balance
	Fees              state.Balance
	Active            bool
	Longitude         int64
	Latitude          int64
}
```
there is not even a single slice and map if you check carefully.
please don't have hard coded rules for return types, methods and arguments.
you should apply what best fits the code, this will not have much impact from the optimisation you're thinking.
we should not have hard and fast rules @samtholiya. please start thinking from what fits best in a given scenario.

it wouldn't really matter much if you use return as value or pointer much
but copies of these big structs would be impactful. and moreover we have no point making copies..
Please go ahead with profiling and benchmark to show the improvement coming from your suggestions.

Reply:
<img width="1220" alt="image" src="https://user-images.githubusercontent.com/11261548/155794964-1a01b9a0-ce4f-438a-97f1-c325c1b17fe9.png">

More processing is required because of the garbage collectors...

Reply:
you're doing benchmarking wrongly to what we're discussing.
try doing benchmark for func calling these.
Take below a very simple example:
```
package benchmark

type X struct {
	a string
	b int
	c int64
	d float64
	e string
	f int
	g string
	h float64
}

func returnStruct() X {
	return X{
		a: "a",
		b: 0,
		c: 0,
		d: 0,
		e: "e",
		f: 0,
		g: "g",
		h: 0,
	}
}

func returnStructPointer() *X {
	return &X{
		a: "a",
		b: 0,
		c: 0,
		d: 0,
		e: "e",
		f: 0,
		g: "g",
		h: 0,
	}
}

func fetchStruct() {
	x := returnStruct()
	transformStruct(x)
}

func fetchStructPointer() {
	x := returnStructPointer()
	transformStructPointer(x)
}

func transformStruct(x X)         {}
func transformStructPointer(x *X) {}
```

I benchmarked `fetchStruct()` and `fetchStructPointer` and these are results:
```
~/GolandProjects/testProject/benchmark > go test -benchmem -run ^$ -bench .                                                                                                       4s
goos: darwin
goarch: arm64
pkg: testProject/benchmark
BenchmarkReturnStruct-8                 286885963                4.155 ns/op           0 B/op          0 allocs/op
BenchmarkReturnStructPointer-8          1000000000               0.3195 ns/op          0 B/op          0 allocs/op
PASS
ok      testProject/benchmark   2.707s
```
creating copies are expensive @samtholiya.
way more expensive than gc calls. and FYI gc trigger can be controlled by you also at any point using GOGC

Reply:
No that is actually the wrong way because:
1. you did not add benchtime. The reason I added bechtime of 10m is so that I can get garbage collector metrics considered in the benchmark. That is the main concern. (heap allocation and garbage collection). Also the data your struct have is not equal to miner struct because more data on struct the faster the heap will be filled.
2. with your code you can not add benchtime because there is no io and therefore the benchmark will freeze if you added benchtime of 10 minutes. because the goroutine will not give up the process (at least happens on my dumb intel mac). 
3. Refer to this go conference that was conducted in Singapore: https://www.youtube.com/watch?v=ZMZpH4yT7M0
4. Refer to this article if you don't have time for the above conference: https://medium.com/@vCabbage/go-are-pointers-a-performance-optimization-a95840d3ef85 and https://www.reddit.com/r/golang/comments/kfd5rr/when_should_i_return_a_pointer_vs_value_when/ 
(not the best explanation)
5. Just like in C++ you can write code that can lead to memory leakage even though it allows for that syntax. The language syntax is not the reason you code in any way to just compile. Golang != Java where you depend on garbage collector only and Golang != C++ where weak pointers exist.
6. In my code the entire logic was common the only change was the one we wanted to inspect. I understand not the best way to benchmark but it was the best way to benchmark considering the fact how golang behaves with goroutine.
so common algo requires common compute cycle
only uncommon algo requires different compute cycle 
so you can see the benchmark showing the diff

Reply:
not sure why ur benchmark is hanging,
but again not sure what u r benchmarking, but I'm discussing the benchmarking of return types and the impact it brings and that's what I have written.

Please take my code, add as much data to struct you want.
that is the bare minimum we want to test, right?

I hope that'll answer the return type(ref vs value) changes that is concerning?

Reply:
All apart!
anyhow I'm not able to run your benchmark as some db connection issue, but I don't see a drastic difference from what your benchmarks are resulting. for you to do deeper dive, try profiling allocation and cpu and see where ur benchmarking is making few ns difference.

but question we're discussing is clearly shown in my benchmarks, of return type pointer vs value.
Please let me know if it doesn't answer the basic question on where performance is getting impacted.

Reply:
I'm doing benchmarking with different struct size. Something interesting to see here. Posting results for better clarity for us:

struct size 10:
```
BenchmarkStructReturnValue
BenchmarkStructReturnValue-8     	149987712	         7.982 ns/op	       0 B/op	       0 allocs/op
BenchmarkStructReturnPointer
BenchmarkStructReturnPointer-8   	268468488	         4.469 ns/op	       0 B/op	       0 allocs/op
```

struct size 100
```
BenchmarkStructReturnValue
BenchmarkStructReturnValue-8     	24166323	        49.78 ns/op	       0 B/op	       0 allocs/op
BenchmarkStructReturnPointer
BenchmarkStructReturnPointer-8   	35991136	        33.19 ns/op	       0 B/op	       0 allocs/op
```

struct size 1000
```
BenchmarkStructReturnValue
BenchmarkStructReturnValue-8     	 2371482	       503.5 ns/op	       0 B/op	       0 allocs/op
BenchmarkStructReturnPointer
BenchmarkStructReturnPointer-8   	 3650842	       328.8 ns/op	       0 B/op	       0 allocs/op
PASS
```

Reply:
To what u r saying, there are functions where returning by value would make sense too. Yes.
But all these theories are for some scenarios, which you've taken as a rule by heart.

And what you're saying, are not yet proven by any of ur benchmarks 😅.
I've shared with you an explaination - a very simple and given 5 benchmarks to show a very simple, not so complicated, comparison.

Let's take this further if any concrete evidence to your points by benchmarking or profiling.
Else please revert to references :)

Reply:
that is valid proof and a golang thing
and a software engineer fundamental "HEAP is not good"
Give me one reference where they say heap is better than stack allocation in long term for performance.
I think the entire golang community is talking from experience and not something as a theory.

I think you need to check the references.
They are from golang community people who work on golang 

Also if you leave golang heap is never a good thing
I hope you read college books as well. And there are many examples

Your program works if I run benchmark directly but that does not account for garbage collection. For that to happen. you have to run a benchmark for enough time so that the heap doubles.

The world is not insane to say that.

Reply:
Onefootball is a company whose entire infrastructure is based on golang.
Also if you see all the functions they have avoided things going to heap.
Even gorm does the same.
So I don't know whom you are following.|
Knowing syntax != knowing language

Reply:
That's great @samtholiya 
Let me go through the above references.

@samtholiya meanwhile can u try having a benchmark written without external db connections and gc triggers as u want. That would help understand our discussion of return types (pointer vs value) for big struct.
Would real help both of us.
Thanks

Reply:
An interesting point to note in ur benchmark screenshot which I'm not able to see difference is.
The benchmark benchmem just captures heap allocation. If we are comparing the heap vs stack on both benchmark, then ur 1 benchmark should have more heap allocation than other. But I see equal in both cases.

Please give me a right benchmark for ur point.
That's the reason I'm saying I haven't got a right benchmark from u:
- db connections should not be part of the benchmark we're discussing
- I'm expecting to see difference in heap allocation of both benchmark
- would be great to also see gc triggers comparison for both scenarios

Reply:
Also benchmark did not hang for you because you did not use benchtime

Reply:
@samtholiya I was able to run even with benchtime, not sure why benchtime will block something.

That being said, don't try benchmark of miner. We don't need db and all to benchmark.

What you need to do to bechmark is:
- benchmark taking just general functions. Big structs and try returning pointer and value.
- please note important aspect here is big struct.
- make sure you don't have any db or external connections in functions. That is not needed for our discussion.

What I'm expecting is:
- time difference.
- allocation differences
Please feel free to use benchtime and other args that makes it helpful.
Also what would be great is to profile gc triggers for both the return types.
Also using go tools you can find out what goes to heap and stack. I guess -m. You can try that also.

All the above points will help us answer what is needed.

Let me know if any doubts.

Reply:
if you see i have changed the code for the benchmark. It does not include DB anymore.

Reply:
also, the reason for bench time is more of a goroutine thing. it is because of the way goroutines are scheduled